src/server/registry/manifest.go

```diff
@@ -49,11 +62,66 @@ func getManifest(w http.ResponseWriter, req *http.Request) {
 		req.URL.RawPath = req.URL.EscapedPath()
 	}
 
-	recorder := lib.NewResponseRecorder(w)
-	proxy.ServeHTTP(recorder, req)
+	// if etag match, we can earlier return and no need to proxy to distribution
+	// as we have stored digest in database.
+	// https://github.com/distribution/distribution/blob/c202b9b0d7b79a67337dec8e1f1bafb1c7095315/registry/handlers/manifests.go#L135
+	if etagMatch(req, art.Digest) {
+		w.WriteHeader(http.StatusNotModified)
+		return
+	}
+
+	buffer := lib.NewResponseBuffer(w)
+	// whether get manifest from cache
+	fromCache := false
+	// whether need to write-back cache
+	wbCache := false
+	// check cache
+	if config.CacheEnabled() {
+		manifest, err := pkg.ManifestMgr.Get(req.Context(), art.Digest)
+		if err == nil {
+			fromCache = true
+			// write header
+			buffer.Header().Set("Content-Length", fmt.Sprintf("%d", len(manifest)))
+			buffer.Header().Set("Content-Type", art.ManifestMediaType)
+			buffer.Header().Set("Docker-Distribution-Api-Version", "registry/2.0")
+			buffer.Header().Set("Docker-Content-Digest", art.Digest)
+			buffer.Header().Set("Etag", fmt.Sprintf(`"%s"`, art.Digest))
+			buffer.WriteHeader(http.StatusOK)
+			// write data from cache, no need to write body if is head request
+			if req.Method == http.MethodGet {
+				buffer.Write(manifest)
+			}
+		} else {
+			log.Warningf("failed to get manifest from cache, error: %v", err)
+			// only write cache when request is GET because HEAD request resp
+			// body is empty.
+			if req.Method == http.MethodGet {
+				wbCache = true
+			}
+		}
+	}
+	// proxy to registry if not from cache
+	if !fromCache {
+		proxy.ServeHTTP(buffer, req)
+	}
+	// flush data
+	if _, err = buffer.Flush(); err != nil {
+		log.Errorf("failed to flush: %v", err)
+		return
+	}
+	// return if not success
+	if !buffer.Success() {
+		return
+	}
+	// write back manifest to cache if needed
+	if wbCache {
```

Comment:
We should exclude the case when the HTTP code is not 200, such as 206. buffer.Success() is not enough.

Reply:
As described in distribution spec, the success HTTP code is 200, ref: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints